File: modules/lut-processor/ios/CubeLutParser.swift
import Foundation

/// Minimal Iridas/Adobe .cube 3D LUT (ASCII) parser: LUT_3D_SIZE, optional DOMAIN, then N^3 R G B lines.
/// File order: red fastest, then green, then blue, mapped to 3D texels (R→x, G→y, B→z).
struct CubeLutData {
  let size: Int
  var domainMin: SIMD3<Float> = .init(0, 0, 0)
  var domainMax: SIMD3<Float> = .init(1, 1, 1)
  /// RGBA8 storage: R-fastest in each slice, layout matches glTexImage3D / Metal 3D (x, y, z) = (r, g, b)
  var rgba8: [UInt8]
}

enum CubeLutError: Error, CustomStringConvertible {
  case fileReadFailed
  case missingSize
  case invalidSize(Int)
  case wrongLineCount(expected: Int, got: Int)
  case badDataLine(String)

  var description: String {
    switch self {
    case .fileReadFailed: return "Could not read .cube file"
    case .missingSize: return "Missing LUT_3D_SIZE in .cube"
    case .invalidSize(let n): return "Invalid LUT_3D_SIZE: \(n)"
    case .wrongLineCount(let e, let g): return "Expected \(e) RGB lines in .cube, got \(g)"
    case .badDataLine(let s): return "Invalid .cube data line: \(s)"
    }
  }
}

enum CubeLutParser {
  static func parse(filePath: String) throws -> CubeLutData {
    let url = URL(fileURLWithPath: filePath)
    let text: String
    do {
      text = try String(contentsOf: url, encoding: .utf8)
    } catch {
      throw CubeLutError.fileReadFailed
    }
    return try parse(text: text)
  }

  static func parse(text: String) throws -> CubeLutData {
    var size = 0
    var domainMin = SIMD3<Float>(0, 0, 0)
    var domainMax = SIMD3<Float>(1, 1, 1)
    var dataLines: [String] = []
    for raw in text.split(whereSeparator: \.isNewline) {
      let line = String(raw).trimmingCharacters(in: .whitespaces)
      if line.isEmpty || line.hasPrefix("#") { continue }
      if line.uppercased().hasPrefix("TITLE") { continue }
      if line.uppercased().hasPrefix("LUT_3D_SIZE") {
        let parts = line.split { $0.isWhitespace }
        guard let last = parts.last, let n = Int(last), n >= 2, n <= 256 else {
          throw CubeLutError.missingSize
        }
        size = n
        continue
      }
      if line.uppercased().hasPrefix("DOMAIN_MIN") {
        domainMin = try parseDomainComponents(line, header: "DOMAIN_MIN")
        continue
      }
      if line.uppercased().hasPrefix("DOMAIN_MAX") {
        domainMax = try parseDomainComponents(line, header: "DOMAIN_MAX")
        continue
      }
      if size > 0 {
        dataLines.append(line)
      }
    }
    guard size > 0 else { throw CubeLutError.missingSize }
    let expected = size * size * size
    guard dataLines.count == expected else {
      throw CubeLutError.wrongLineCount(expected: expected, got: dataLines.count)
    }
    var rgba = [UInt8](repeating: 0, count: expected * 4)
    for (i, line) in dataLines.enumerated() {
      let rgb = try parseRGBLine(line)
      let n = size
      let r = i % n
      let g = (i / n) % n
      let b = i / (n * n)
      let mem = (r + g * n + b * n * n) * 4
      let cr = min(1, max(0, rgb.r))
      let cg = min(1, max(0, rgb.g))
      let cb = min(1, max(0, rgb.b))
      rgba[mem + 0] = UInt8(min(255, max(0, Int((cr * 255.0).rounded()))))
      rgba[mem + 1] = UInt8(min(255, max(0, Int((cg * 255.0).rounded()))))
      rgba[mem + 2] = UInt8(min(255, max(0, Int((cb * 255.0).rounded()))))
      rgba[mem + 3] = 255
    }
    return CubeLutData(size: size, domainMin: domainMin, domainMax: domainMax, rgba8: rgba)
  }

  private static func parseDomainComponents(
    _ line: String,
    header: String
  ) throws -> SIMD3<Float> {
    let parts = line.split { $0.isWhitespace || $0 == "\t" }.filter { !$0.isEmpty }
    guard parts.count >= 4,
          parts[0].uppercased() == header.uppercased(),
          let a = Float(parts[1]),
          let b = Float(parts[2]),
          let c = Float(parts[3]) else {
      throw CubeLutError.badDataLine(line)
    }
    return SIMD3(a, b, c)
  }

  private static func parseRGBLine(_ line: String) throws -> (r: Float, g: Float, b: Float) {
    let parts = line.split { $0.isWhitespace || $0 == "\t" }.filter { !$0.isEmpty }
    guard parts.count >= 3,
          let r = Float(parts[0]), let g = Float(parts[1]), let b = Float(parts[2]) else {
      throw CubeLutError.badDataLine(line)
    }
    return (r, g, b)
  }
}

File: modules/lut-processor/android/src/main/java/expo/modules/lutprocessor/CubeLutParser.kt
package expo.modules.lutprocessor

import java.io.File
import kotlin.text.Charsets

data class CubeLutData(
  val size: Int,
  val domainMin: FloatArray = floatArrayOf(0f, 0f, 0f),
  val domainMax: FloatArray = floatArrayOf(1f, 1f, 1f),
  /** Red-fastest .cube rows, layout: offset = (r + g*size + b*size*size) * 4, RGBA8 */
  val rgba8: ByteArray,
) {
  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false
    other as CubeLutData
    if (size != other.size) return false
    if (!domainMin.contentEquals(other.domainMin)) return false
    if (!domainMax.contentEquals(other.domainMax)) return false
    if (!rgba8.contentEquals(other.rgba8)) return false
    return true
  }

  override fun hashCode(): Int {
    var result = size
    result = 31 * result + domainMin.contentHashCode()
    result = 31 * result + domainMax.contentHashCode()
    result = 31 * result + rgba8.contentHashCode()
    return result
  }
}

object CubeLutParser {
  fun parseFile(path: String): CubeLutData {
    val text = File(path).readText(Charsets.UTF_8)
    return parse(text)
  }

  fun parse(text: String): CubeLutData {
    var size = 0
    var domainMin = floatArrayOf(0f, 0f, 0f)
    var domainMax = floatArrayOf(1f, 1f, 1f)
    val dataLines = ArrayList<String>(4096)
    for (raw in text.split('\n', '\r')) {
      val line = raw.trim()
      if (line.isEmpty() || line.startsWith("#")) continue
      if (line.length >= 5 && line.substring(0, 5).equals("TITLE", ignoreCase = true)) continue
      if (line.length >= 12 && line.substring(0, 12).equals("LUT_3D_SIZE", ignoreCase = true)) {
        val n = line.split(Regex("\\s+")).lastOrNull()?.toIntOrNull()
        requireNotNull(n) { "Invalid LUT_3D_SIZE" }
        require(n in 2..256) { "Invalid LUT_3D_SIZE: $n" }
        size = n
        continue
      }
      if (line.length >= 9 && line.substring(0, 9).equals("DOMAIN_MIN", ignoreCase = true)) {
        domainMin = parseThreeFloats(line, "DOMAIN_MIN")
        continue
      }
      if (line.length >= 9 && line.substring(0, 9).equals("DOMAIN_MAX", ignoreCase = true)) {
        domainMax = parseThreeFloats(line, "DOMAIN_MAX")
        continue
      }
      if (size > 0) {
        dataLines.add(line)
      }
    }
    require(size > 0) { "Missing LUT_3D_SIZE in .cube" }
    val expected = size * size * size
    require(dataLines.size == expected) {
      "Expected $expected RGB lines in .cube, got ${dataLines.size}"
    }
    val rgba = ByteArray(expected * 4)
    val n = size
    for (i in 0 until expected) {
      val rgb = parseRgbLine(dataLines[i])
      val r = i % n
      val g = (i / n) % n
      val b = i / (n * n)
      val mem = (r + g * n + b * n * n) * 4
      val cr = rgb[0].coerceIn(0f, 1f)
      val cg = rgb[1].coerceIn(0f, 1f)
      val cb = rgb[2].coerceIn(0f, 1f)
      rgba[mem] = (cr * 255f).toInt().coerceIn(0, 255).toByte()
      rgba[mem + 1] = (cg * 255f).toInt().coerceIn(0, 255).toByte()
      rgba[mem + 2] = (cb * 255f).toInt().coerceIn(0, 255).toByte()
      rgba[mem + 3] = 255.toByte()
    }
    return CubeLutData(
      size = size,
      domainMin = domainMin,
      domainMax = domainMax,
      rgba8 = rgba,
    )
  }

  private fun parseThreeFloats(line: String, header: String): FloatArray {
    val parts = line.split(Regex("\\s+")).filter { it.isNotEmpty() }
    require(parts[0].equals(header, ignoreCase = true)) { line }
    require(parts.size >= 4) { line }
    return floatArrayOf(parts[1].toFloat(), parts[2].toFloat(), parts[3].toFloat())
  }

  private fun parseRgbLine(line: String): FloatArray {
    val parts = line.split(Regex("\\s+|\t+")).filter { it.isNotEmpty() }
    require(parts.size >= 3) { line }
    return floatArrayOf(parts[0].toFloat(), parts[1].toFloat(), parts[2].toFloat())
  }
}
